### PR TITLE
Eliminate throw from array subscript error paths

### PR DESCRIPTION
**Pull request by oerling**
Eliminate throw from array subscript error paths

Throw in frequent error path inside try destroys performance. Fix this
for array subscript. More such fixes are to be expected.

### CODE REVIEW

File: velox/functions/lib/SubscriptUtil.h [hunk 1]
```diff
@@ -187,35 +230,54 @@ class SubscriptImpl : public exec::Subscript {
...
       bool allFailed = false;
       // If index is invalid, capture the error and mark all rows as failed.
-      try {
-        adjustedIndex = adjustIndex(decodedIndices->valueAt<I>(0));
-      } catch (const VeloxRuntimeError&) {
-        throw;
-      } catch (const std::exception& e) {
-        context.setErrors(rows, std::current_exception());
+      bool isZero = false;
+      adjustedIndex = adjustIndex(decodedIndices->valueAt<I>(0), isZero);
```
Comment (xiaoxmeng) [inline velox/functions/lib/SubscriptUtil.h:242]: How about rename isZero to hasZeroSubscriptError?

File: velox/functions/lib/SubscriptUtil.h [hunk 2]
```diff
@@ -187,35 +230,54 @@ class SubscriptImpl : public exec::Subscript {
     auto rawSizes = baseArray->rawSizes();
     auto rawOffsets = baseArray->rawOffsets();
 
+    SelectivityVector errorRows;
+    std::exception_ptr error;
```
Comment (xiaoxmeng) [inline velox/functions/lib/SubscriptUtil.h:234]:
```
// Used to store the first encountered row error during the processing below.
std::exception_ptr error{nullptr};
```

File: velox/functions/lib/SubscriptUtil.h [hunk 3]
```diff
@@ -187,35 +230,54 @@ class SubscriptImpl : public exec::Subscript {
...
         });
       }
     } else {
-      context.applyToSelectedNoThrow(rows, [&](auto row) {
-        auto adjustedIndex = adjustIndex(decodedIndices->valueAt<I>(row));
-        auto elementIndex =
-            getIndex(adjustedIndex, row, rawSizes, rawOffsets, arrayIndices);
+      rows.applyToSelected([&](auto row) {
+        auto originalIndex = decodedIndices->valueAt<I>(row);
+        bool isZero = false;
```
Comment (xiaoxmeng) [inline velox/functions/lib/SubscriptUtil.h:267]: ditto

File: velox/functions/lib/SubscriptUtil.h [hunk 4]
```diff
@@ -187,35 +230,54 @@ class SubscriptImpl : public exec::Subscript {
...
+        auto adjustedIndex = adjustIndex(originalIndex, isZero);
```
Comment (xiaoxmeng) [inline velox/functions/lib/SubscriptUtil.h:268]:
```
const auto adjustedIndex
```

File: velox/functions/lib/SubscriptUtil.h (hunk 1 above)
Comment (xiaoxmeng) [inline velox/functions/lib/SubscriptUtil.h:242]:
```
const vector_size_t adjustedIndex = adjustIndex(decodedIndices->valueAt<I>(0), isZero);
```

File: velox/functions/lib/SubscriptUtil.h [hunk 5]
```diff
@@ -229,6 +291,10 @@ class SubscriptImpl : public exec::Subscript {
       return BaseVector::createNullConstant(
           baseArray->elements()->type(), rows.end(), context.pool());
     }
+    if (error) {
```
Comment (xiaoxmeng) [inline velox/functions/lib/SubscriptUtil.h:294]:
```
if (error != nullptr) {
```

File: velox/functions/lib/SubscriptUtil.h [hunk 6]
```diff
@@ -187,35 +230,54 @@ class SubscriptImpl : public exec::Subscript {
...
+        context.setErrors(rows, zeroSubscriptError());
         allFailed = true;
       }
 
       if (!allFailed) {
-        context.applyToSelectedNoThrow(rows, [&](auto row) {
-          auto elementIndex =
-              getIndex(adjustedIndex, row, rawSizes, rawOffsets, arrayIndices);
+        rows.applyToSelected([&](auto row) {
+          auto elementIndex = getIndex(
```
Comment (xiaoxmeng) [inline velox/functions/lib/SubscriptUtil.h:250]:
```
const auto elementIndex =
```

File: velox/functions/lib/SubscriptUtil.h [hunk 7]
```diff
@@ -295,6 +363,20 @@ class SubscriptImpl : public exec::Subscript {
     return rawOffsets[indices[row]] + index;
   }
 
+  void addError(
+      vector_size_t row,
+      std::exception_ptr newError,
+      std::exception_ptr& error,
+      SelectivityVector& errorRows) const {
+    if (!error) {
```
Comment (xiaoxmeng) [inline velox/functions/lib/SubscriptUtil.h:371]:
```
nit: if (error != nullptr)
```

File: velox/functions/lib/SubscriptUtil.h [hunk 8]
```diff
@@ -295,6 +363,20 @@ class SubscriptImpl : public exec::Subscript {
     return rawOffsets[indices[row]] + index;
   }
 
+  void addError(
+      vector_size_t row,
+      std::exception_ptr newError,
```
Comment (xiaoxmeng) [inline velox/functions/lib/SubscriptUtil.h:368]:
```
const std::exception_ptr& newError,
```

**Review by xiaoxmeng — COMMENTED** [review]
@oerling the change looks good to me. But I am not sure if we return the first encountered error for the failed rows are ok or not. Does this change breaks any existing test or no unit test currently catch this behavior? Thanks!

File: velox/functions/lib/SubscriptUtil.h [hunk 9]
```diff
@@ -295,6 +363,20 @@ class SubscriptImpl : public exec::Subscript {
...
+      std::exception_ptr& error,
+      SelectivityVector& errorRows) const {
+    if (!error) {
+      error = newError;
+    }
+    if (errorRows.size() <= row) {
+      errorRows.resize(row + 1, false);
+    }
+    errorRows.setValid(row, true);
```
Comment (xiaoxmeng) [inline velox/functions/lib/SubscriptUtil.h:377]: Can record the errors for each failed row? Have a chat with Wei and some Meta internal query system might need these distinct error type report for monitoring purpose. Thanks!

File: velox/functions/lib/SubscriptUtil.h [hunk 10]
```diff
@@ -23,6 +23,49 @@
 
 namespace facebook::velox::functions {
 
+namespace {
```
Comment (Yuhta) [inline velox/functions/lib/SubscriptUtil.h:26]: @oerling We should not have anonymous namespace in header file.  Maybe using `detail` namespace instead

**Pull request closed without merging**